experimental/cache_size.h

```diff
@@ -6,47 +6,90 @@
 #include <memory>
 #include <random>
 #include <string>
+#include <assert.h>
 
 // TODO: depending on in which directory the file end up these might need to be
 // changed
 #include "../demos/asm/measurereadlatency.h"
 #include "../demos/instr.h"
 #include "../demos/utils.h"
 
-// Returns a shuffled vector of integers from 0 (inclusive) to n (exclusive).
-// This can be used to randomize access order.
-inline std::vector<size_t> ShuffledRange(size_t n) {
-  std::vector<size_t> numbers;
-  numbers.reserve(n);
-  for (size_t i = 0; i < n; ++i) {
-    numbers.push_back(i);
+
+// A permutation function for 1 byte.
+inline unsigned char PermuteChar(unsigned char x) {
+  // LCG
+  return x * 113 + 100;
```

Comment:
This isn't a linear congruential generator.

In an LCG the output is iteratively fed back into the input to create a pseudorandom sequence. Here, though, we're just passing the numbers 0.._n_-1 into this function expecting to get a permutation of 0.._n_-1 back.

This works because we're generating the finite cyclic group of integers under addition modulo _m_. That cyclic group is generated by any element coprime to the modulus. Here _m_ is 2<sup>8</sup>, so any odd number is coprime. [ref](https://en.wikipedia.org/wiki/Cyclic_group#Integer_and_modular_addition).

We added the 100 to avoid mapping 0 to 0. Unlike an LCG, it doesn't play into the "permutation-ness" at all.